src/Passes/docs/base-profile-transformations.md

```diff
@@ -0,0 +1,233 @@
+
+## Goal
+
+We envision the tool to work as a standalone commandline tool which can either validate or generate a QIR in accordance with a given profile. To validate, one would run:
+
+```language
+qat -p profile.yaml --validate unvalidated-qir.ll
+```
+
+In a similar fashion, generation is performed by adding `--generate` to the commandline:
+
+```language
+qat -p profile.yaml --generate qir.ll > qir-profile.ll
+```
+
+Default behaviour of the tool is that it always validates the generated profile. This behaviour can be disabled by
+
+```language
+qat -p profile.yaml --generate --no-validate qir.ll > qir-profile.ll
+```
+
+# Profile Specification
+
+Every profile is specified through a YAML file which defines an object at the top-level. This object must contain the fields `name` and `displayName`:
+
+```yaml
+name: profile-name
+displayName: Profile Name
+# ...
+```
+
+Additionally, top level also contains the fields `version` and `mode`. The version refers to the QIR version which forms the basis for the specification and `mode` explains how the profile is defined.
+The final two top level fields are lists named `specification` and `generation`. These contains the specification and generation producedure, respectively.
+
+## Specification
+
+The default `mode` of specification is by `feature` which means that specification describes the feature set available. Alternatively, one can specify a profile by `limitation`. As an example profile for
+
+```yaml
+name: profile-name
+displayName: Profile Name
+version: 1.0
+mode: feature
+specification:
+  functions:
+    - __quantum__qis__toffoli__body
+    - __quantum__qis__cnot__body
+    - __quantum__qis__cz__body
+    - __quantum__qis__h__body
+    - __quantum__qis__mz__body
+    - __quantum__qis__reset__body
+    - __quantum__qis__rx__body
+    - __quantum__qis__ry__body
+    - __quantum__qis__rz__body
+    - __quantum__qis__s__body
+    - __quantum__qis__s_adj
+    - __quantum__qis__t__body
+    - __quantum__qis__t__adj
+    - __quantum__qis__x__body
+    - __quantum__qis__y__body
+    - __quantum__qis__z__body
+  instructions:
+    - call
+    - br
+    - ret
+    - inttoptr
+# ...
+```
+
+This specification describes that 16 quantum instructions are available and 4 classical operations of the full QIR spec. Contrary, a specification by `limitation` could be as follows:
+
+```yaml
+name: profile-name
+displayName: Profile Name
+version: 1.0
+mode: limitation
+specification:
+  functions:
+    - __quantum__rt__array_update_alias_count
+    - __quantum__rt__array_update_reference_count
+  instructions:
+    - br
+# ...
+```
+
+This profile specifies a system that does not allow reference and alias counting and neither have support for branching, but otherwise has the full QIR vesion 1.0 available.
+
+## Generation specification
```

Comment:
It seems to me that there is some partial redundancy between the profile specification and the generation. I take it that the profile specification does not actually impact at all which passes are run? Is it correct that the profile specification is only used for validation where as the generation specification is used only for generation?

Reply:
That is the thought. Specification is essentially an analysis whereas generation is a transformation.  Specification has a predetermined set of passes it will run and the specification will tell these passes what is allowed and not allowed. This is mostly due to that I could not think of any case where you needed to truly compose passes to define the profile. 

Generation, on the other hand, is a far more complex and require certain level of flexibility and configurability since it transforms (generates) the IR.